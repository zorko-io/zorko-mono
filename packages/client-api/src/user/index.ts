export * from './user.api.client';
export * from './user.axios.api.client';
