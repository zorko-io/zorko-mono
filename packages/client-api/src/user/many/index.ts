export * from './user.many.api.client'
export * from './user.axios.many.api.client'
