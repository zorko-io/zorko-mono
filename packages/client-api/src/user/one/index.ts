export * from './user.one.api.client'
export * from './user.axios.one.api.client'
