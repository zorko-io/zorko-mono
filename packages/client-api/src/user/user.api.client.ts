import { AuthApiClient } from '../auth';
import { RemoteUserApi } from '@zorko/remote-api';

export interface UserApiClient extends AuthApiClient, RemoteUserApi {}
