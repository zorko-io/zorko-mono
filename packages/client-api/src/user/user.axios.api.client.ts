import { AxiosResponse } from 'axios';
import { AuthAxiosApiClient } from '../auth';
import { UserDto } from '@zorko/dto';
import { UserApiClient } from './user.api.client';
import {
  CreateUserCollectionParams,
  CreateUserParams, DeleteUserCollectionParams, DeleteUserParams,
  ReadUserCollectionParams,
  ReadUserParams,
  UpdateUserCollectionParams, UpdateUserParams,
} from '@zorko/remote-api';

export class UserAxiosApiClient extends AuthAxiosApiClient implements UserApiClient  {

  async findMany(readParams?: ReadUserCollectionParams): Promise<UpdateUserCollectionParams> {
    const response: AxiosResponse<UpdateUserCollectionParams> = await this.http.get('/users');
    return response.data
  }

  async createOne(user: CreateUserParams): Promise<string> {
    const response: AxiosResponse<string> = await this.http.post('/users', user);
    return response.data;
  }

  async findOne(readParams: ReadUserParams): Promise<UpdateUserParams> {
    const { id }  = readParams;
    const response: AxiosResponse<UserDto> = await this.http.get(`/users/${id}`);
    return response.data;
  }

  async createMany(createParams: CreateUserCollectionParams): Promise<string> {
    return '123'
  }

  async removeMany(deleteParams: DeleteUserCollectionParams): Promise<string> {
    const response: AxiosResponse<string> = await this.http.delete('/users');
    return response.data
  }

  async removeOne(deleteParams: DeleteUserParams): Promise<void> {
    const { id } = deleteParams;
    await this.http.delete(`/users/${id}`);
  }

  updateMany(updateParams: UpdateUserCollectionParams): Promise<ReadUserCollectionParams> {
    return undefined;
  }

  updateOne(updateParams: UpdateUserParams): Promise<ReadUserParams> {
    return undefined;
  }
}

