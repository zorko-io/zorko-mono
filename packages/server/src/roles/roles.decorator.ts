import { SetMetadata} from '@nestjs/common';
import { RolesEnum } from '@zorko/dto';

export const RolesMetadataKey = 'roles';

export const Roles = (...roles: RolesEnum[]) => SetMetadata(RolesMetadataKey, roles);
