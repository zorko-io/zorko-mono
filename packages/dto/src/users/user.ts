import { RolesEnum } from '../roles';

export interface User {
  id?: any;
  email: string;
  password?: string;
  hashPassword?: string;

  roles?: RolesEnum[];
}
