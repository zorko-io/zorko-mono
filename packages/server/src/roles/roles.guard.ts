import { CanActivate, ExecutionContext, Injectable } from '@nestjs/common';
import { Observable } from 'rxjs';
import { Reflector } from '@nestjs/core';
import { RolesEnum } from '@zorko/dto';
import { RolesMetadataKey } from './roles.decorator';
import { ConfigService } from '../config/config.service';

@Injectable()
export class RolesGuard implements CanActivate {
  constructor(
    private readonly reflector: Reflector,
    private readonly configService: ConfigService
  ) {}

  canActivate(context: ExecutionContext): boolean | Promise<boolean> | Observable<boolean> {
    if (!this.configService.isAuthEnabled) {
      return true;
    }

    const roles = this.reflector.get<RolesEnum[]>(RolesMetadataKey, context.getHandler());
    if (!roles) {
      return true;
    }

    const request = context.switchToHttp().getRequest();
    const user = request.user;
    const hasRole = () => user.roles.some((role) => roles.includes(role));

    return user && user.roles && hasRole();
  }
}
