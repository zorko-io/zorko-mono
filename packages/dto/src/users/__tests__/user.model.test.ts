import { UserModel } from '../user.model';
import { RolesEnum } from '../../roles';
import { User } from '../user';

describe('User', () => {

  let user: UserModel;
  let actual: User;

  beforeEach(() => {
    user = new UserModel('test@email.com', 'bobobobo');
  });

  it('creates user with defaults', () => {
      expect(user.setId('fdsfsdfsdf').toDTO()).toMatchSnapshot();
  });

  it('updates id', () => {
    expect(user.setId('fdsfsdfsdf').toDTO()).toMatchSnapshot();
    expect(user.getId()).toEqual('fdsfsdfsdf');
  });

  it('updates email', () => {
    expect(user.setEmail('test@email.com').toDTO()).toMatchSnapshot();
    expect(user.getEmail()).toEqual('test@email.com');
  });

  it('updates password', () => {
    expect(user.setPassword('39393kfkfkf').toDTO()).toMatchSnapshot();
    expect(user.getPassword()).toEqual('39393kfkfkf');
  });

  it('updates hashPassword', () => {
   user.setHashPassword('bkbjf848484');

    expect(user.getHashPassword()).toEqual('bkbjf848484');
    expect(user.getPassword()).toBeUndefined();
  });

  it('updates roles', () => {
    expect(user.setRoles([RolesEnum.User, RolesEnum.Admin]).toDTO()).toMatchSnapshot();
    expect(user.getRoles()).toEqual([RolesEnum.User, RolesEnum.Admin]);
  });

  it('fails on wrong email', () => {
    user = user.setEmail('fdfddf');

    try {
      user.toDTO()
    } catch (e) {
      expect(e).toMatchSnapshot();
    }

  })

});
