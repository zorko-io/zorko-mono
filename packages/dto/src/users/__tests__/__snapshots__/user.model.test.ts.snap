// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User creates user with defaults 1`] = `
Object {
  "email": "test@email.com",
  "id": "fdsfsdfsdf",
  "password": "bobobobo",
  "roles": undefined,
}
`;

exports[`User fails on wrong email 1`] = `[ValidationError: email must be a valid email]`;

exports[`User updates email 1`] = `
Object {
  "email": "test@email.com",
  "id": undefined,
  "password": "bobobobo",
  "roles": undefined,
}
`;

exports[`User updates id 1`] = `
Object {
  "email": "test@email.com",
  "id": "fdsfsdfsdf",
  "password": "bobobobo",
  "roles": undefined,
}
`;

exports[`User updates password 1`] = `
Object {
  "email": "test@email.com",
  "id": undefined,
  "password": "39393kfkfkf",
  "roles": undefined,
}
`;

exports[`User updates roles 1`] = `
Object {
  "email": "test@email.com",
  "id": undefined,
  "password": "bobobobo",
  "roles": Array [
    "user",
    "admin",
  ],
}
`;
